lib/spack/spack/bootstrap.py

```diff
@@ -837,3 +838,156 @@ def ensure_flake8_in_path_or_raise():
     """Ensure that flake8 is in the PATH or raise."""
     executable, root_spec = 'flake8', flake8_root_spec()
     return ensure_executables_in_path_or_raise([executable], abstract_spec=root_spec)
+
+
+def _missing_executable(name, purpose):
+    """Message to be printed if an executable is not found"""
+    msg = 'MISSING: "{0}" executable (required to {1})'
+    return msg.format(name, purpose)
+
+
+def _system_executable(exes, msg):
+    """Query availability of executables to be found on the system"""
```

Comment:
Could you update the docstring and name of this function to contrast it more directly with `_required_executable`? e.g. if you call this `_required_system_executable` and `_required_executable` can mention that the exe can be found on the system path or in the bootstrap store.

Also I'm assuming that for example if `unzip` were found in the bootstrap store, that we wouldn't be able to use it. If that's not true though, I wonder what the need is for a separate `_system_executable` and `_required_executable` function.

Reply:
I think in the long run, we should go through this list and make sure that everything that we bootstrap everything in here that we can.

Reply:
@becker33 Yes, that is the idea going forward.

Reply:
@scheibelp I think it's fixed in 830c4f3

Reply:
The docstrings look great. I should have been explicit that I had a question though:

> Also I'm assuming that for example if unzip were found in the bootstrap store, that we wouldn't be able to use it. If that's not true though, I wonder what the need is for a separate _system_executable and _required_executable function.

If an executable happens to be found in the bootstrap directory but not in the system directories, is it the case now that we wouldn't be able to use it? If we can still use it, then I'm not sure if we need a distinct `_required_system_executable` function (IMO `_required_executable` would be sufficient).

Reply:
I kept the separation since we are trying to bootstrap only a restricted number of specs at the moment (`clingo`, `gnupg` and the Python modules needed by style checks) and I am using `_required_executable` only for them. I think in the future more dependencies will move from being system requirements to be bootstrappable dependencies, though I am not sure if we can move all of them - `tar` for instance is needed to manage buildcache archives and source builds.

Reply:
So to answer more directly, an executable that we don't try to bootstrap explicitly will not be found in the bootstrap store and we need a few lines of code at calling site to be able to use it. For `gnupg` this is:

https://github.com/spack/spack/blob/db69a291d4937c230517a6f049010737c98ac1d8/lib/spack/spack/util/gpg.py#L63-L65

Reply:
> an executable that we don't try to bootstrap explicitly will not be found in the bootstrap store and we need a few lines of code at calling site to be able to use it. 

Thanks for the example, if executables in the bootstrap store are not added to `PATH` etc., then it makes sense to have two functions (I'm separately curious why that distinction exists, but it's not something that needs to be dealt with in this PR if at all).